packages/runed/src/lib/utilities/NetworkStatus/NetworkStatus.svelte.ts

```diff
@@ -0,0 +1,166 @@
+import { untrack } from "svelte";
+import { browser } from "$lib/internal/utils/browser.js";
+import { addEventListener } from "$lib/internal/utils/event.js";
+import { IsSupported } from "$lib/utilities/index.js";
+
+/**
+ * @desc The `NetworkInformation` interface of the Network Information API
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation
+ */
+type NetworkInformation = {
+	readonly downlink: number;
+	readonly downlinkMax: number;
+	readonly effectiveType: "slow-2g" | "2g" | "3g" | "4g";
+	readonly rtt: number;
+	readonly saveData: boolean;
+	readonly type:
+		| "bluetooth"
+		| "cellular"
+		| "ethernet"
+		| "none"
+		| "wifi"
+		| "wimax"
+		| "other"
+		| "unknown";
+} & EventTarget;
+
+type NavigatorWithConnection = Navigator & { connection: NetworkInformation };
+
+/**
+ * Tracks the state of browser's network connection.
+ */
+export class NetworkStatus {
+	#isNetworkStatusSupported = new IsSupported(() => browser && "navigator" in window);
+	#navigator?: Navigator = $derived(
+		this.#isNetworkStatusSupported.current ? window.navigator : undefined
+	);
+	#connection?: NetworkInformation = $derived(
+		this.#navigator && "connection" in this.#navigator
+			? (this.#navigator as NavigatorWithConnection).connection
+			: undefined
+	);
+
+	#isSupported = $derived(this.#isNetworkStatusSupported.current);
+	#online: boolean = $state(false);
+	#updatedAt: Date = $state(new Date());
+	#downlink?: NetworkInformation["downlink"] = $state();
+	#downlinkMax?: NetworkInformation["downlinkMax"] = $state();
+	#effectiveType?: NetworkInformation["effectiveType"] = $state();
+	#rtt?: NetworkInformation["rtt"] = $state();
+	#saveData?: NetworkInformation["saveData"] = $state();
+	#type?: NetworkInformation["type"] = $state();
+
+	constructor() {
+		$effect(() => {
+			untrack(() => {
+				this.#updateStatus();
+			});
+
+			const callbacks: VoidFunction[] = [];
+
+			// The connection event handler also manages online and offline states.
+			if (this.#connection) {
+				callbacks.push(
+					addEventListener(this.#connection, "change", this.#updateStatus, { passive: true })
+				);
+			} else {
+				callbacks.push(addEventListener(window, "online", this.#updateStatus, { passive: true }));
+				callbacks.push(addEventListener(window, "offline", this.#updateStatus, { passive: true }));
+			}
+
+			return () => {
+				callbacks.forEach((c) => c());
+			};
+		});
+	}
+
+	#updateStatus = () => {
+		if (this.#navigator) {
+			this.#online = this.#navigator.onLine;
+			this.#updatedAt = new Date();
+			if (this.#connection) {
+				this.#downlink = this.#connection.downlink;
+				this.#downlinkMax = this.#connection.downlinkMax;
+				this.#effectiveType = this.#connection.effectiveType;
+				this.#rtt = this.#connection.rtt;
+				this.#saveData = this.#connection.saveData;
+				this.#type = this.#connection.type;
+			}
+		}
```

Comment:
nitpick(optional): This is purely personal coding style for early returns, but it's 100% fine as is too 😅

```suggestion
		if (!this.#navigator) return;

		this.#online = this.#navigator.onLine;
		this.#updatedAt = new Date();

		if (!this.#connection) return;

		this.#downlink = this.#connection.downlink;
		this.#downlinkMax = this.#connection.downlinkMax;
		this.#effectiveType = this.#connection.effectiveType;
		this.#rtt = this.#connection.rtt;
		this.#saveData = this.#connection.saveData;
		this.#type = this.#connection.type;
```